File: client/mia_companion/lib/widgets/mia_presence_row.dart
import 'dart:math' as math;

import 'package:flutter/material.dart';

import '../theme/mia_theme.dart';
import 'mia_avatar.dart';
import 'message_bubble.dart';

enum MiaPresenceKind { typing, recording }

/// WhatsApp-style presence: Mia's avatar + compact activity bubble.
class MiaPresenceRow extends StatelessWidget {
  const MiaPresenceRow({
    super.key,
    required this.kind,
    this.compactTop = false,
  });

  final MiaPresenceKind kind;
  final bool compactTop;

  static const double _avatarSize = 28;
  static const double _bubbleMinHeight = 36;

  @override
  Widget build(BuildContext context) {
    return Padding(
      padding: EdgeInsets.only(
        top: compactTop
            ? MessageBubble.gapSameSender
            : MessageBubble.gapNewSender,
        bottom: 2,
      ),
      child: Row(
        crossAxisAlignment: CrossAxisAlignment.end,
        children: [
          MiaAvatar(size: _avatarSize),
          const SizedBox(width: 8),
          _PresenceBubble(kind: kind),
        ],
      ),
    );
  }
}

class _PresenceBubble extends StatelessWidget {
  const _PresenceBubble({required this.kind});

  final MiaPresenceKind kind;

  @override
  Widget build(BuildContext context) {
    return Container(
      constraints: const BoxConstraints(minHeight: MiaPresenceRow._bubbleMinHeight),
      padding: EdgeInsets.symmetric(
        horizontal: kind == MiaPresenceKind.typing ? 14 : 16,
        vertical: kind == MiaPresenceKind.typing ? 10 : 9,
      ),
      decoration: BoxDecoration(
        color: MiaColors.miaBubble,
        borderRadius: MiaTheme.bubbleShape(isUser: false),
        boxShadow: [
          BoxShadow(
            color: Colors.black.withValues(alpha: 0.04),
            blurRadius: 6,
            offset: const Offset(0, 1),
          ),
        ],
      ),
      child: kind == MiaPresenceKind.typing
          ? const _WhatsAppTypingDots()
          : const _WhatsAppRecordingMic(),
    );
  }
}

/// Three dots that bounce in sequence (WhatsApp typing indicator).
class _WhatsAppTypingDots extends StatefulWidget {
  const _WhatsAppTypingDots();

  @override
  State<_WhatsAppTypingDots> createState() => _WhatsAppTypingDotsState();
}

class _WhatsAppTypingDotsState extends State<_WhatsAppTypingDots>
    with SingleTickerProviderStateMixin {
  late final AnimationController _controller;

  static const _dotSize = 6.0;
  static const _dotSpacing = 4.0;
  static const _bounce = 4.0;

  @override
  void initState() {
    super.initState();
    _controller = AnimationController(
      vsync: this,
      duration: const Duration(milliseconds: 1200),
    )..repeat();
  }

  @override
  void dispose() {
    _controller.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return AnimatedBuilder(
      animation: _controller,
      builder: (context, _) {
        return SizedBox(
          height: _dotSize + _bounce,
          child: Row(
            mainAxisSize: MainAxisSize.min,
            crossAxisAlignment: CrossAxisAlignment.end,
            children: List.generate(3, (i) {
              final phase = (_controller.value * 2 * math.pi) - (i * 0.75);
              final y = math.sin(phase) * _bounce;
              return Padding(
                padding: EdgeInsets.only(left: i == 0 ? 0 : _dotSpacing),
                child: Transform.translate(
                  offset: Offset(0, -y),
                  child: Container(
                    width: _dotSize,
                    height: _dotSize,
                    decoration: BoxDecoration(
                      color: MiaColors.textMuted.withValues(
                        alpha: 0.55 + 0.35 * ((math.sin(phase) + 1) / 2),
                      ),
                      shape: BoxShape.circle,
                    ),
                  ),
                ),
              );
            }),
          ),
        );
      },
    );
  }
}

/// Mic icon pulse while Mia is "recording audio" (WhatsApp-style).
class _WhatsAppRecordingMic extends StatefulWidget {
  const _WhatsAppRecordingMic();

  @override
  State<_WhatsAppRecordingMic> createState() => _WhatsAppRecordingMicState();
}

class _WhatsAppRecordingMicState extends State<_WhatsAppRecordingMic>
    with SingleTickerProviderStateMixin {
  late final AnimationController _controller;
  late final Animation<double> _pulse;

  static const _micColor = Color(0xFFE54D42);

  @override
  void initState() {
    super.initState();
    _controller = AnimationController(
      vsync: this,
      duration: const Duration(milliseconds: 900),
    )..repeat(reverse: true);
    _pulse = Tween(begin: 0.75, end: 1.0).animate(
      CurvedAnimation(parent: _controller, curve: Curves.easeInOut),
    );
  }

  @override
  void dispose() {
    _controller.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return AnimatedBuilder(
      animation: _pulse,
      builder: (context, child) {
        return Opacity(
          opacity: _pulse.value,
          child: Transform.scale(
            scale: 0.92 + (_pulse.value * 0.08),
            child: child,
          ),
        );
      },
      child: const Icon(
        Icons.mic_rounded,
        size: 18,
        color: _micColor,
      ),
    );
  }
}

File: client/mia_android/app/src/main/kotlin/com/mia/companion/ui/components/MiaPresenceRow.kt
package com.mia.companion.ui.components

import androidx.compose.animation.core.LinearEasing
import androidx.compose.animation.core.RepeatMode
import androidx.compose.animation.core.animateFloat
import androidx.compose.animation.core.infiniteRepeatable
import androidx.compose.animation.core.rememberInfiniteTransition
import androidx.compose.animation.core.tween
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.ui.draw.clip
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.rounded.Mic
import androidx.compose.material3.Icon
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.alpha
import androidx.compose.ui.draw.scale
import androidx.compose.ui.draw.shadow
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.unit.dp
import com.mia.companion.ui.theme.MiaColors
import com.mia.companion.ui.theme.bubbleShape
import kotlin.math.sin

enum class MiaPresenceKind { Typing, Recording }

@Composable
fun MiaPresenceRow(
    kind: MiaPresenceKind,
    compactTop: Boolean,
    modifier: Modifier = Modifier,
) {
    val topPad = if (compactTop) 4.dp else 16.dp
    Row(
        modifier = modifier
            .padding(top = topPad, bottom = 2.dp),
        verticalAlignment = Alignment.Bottom,
    ) {
        MiaAvatar(size = 28.dp)
        Box(Modifier.width(8.dp))
        PresenceBubble(kind)
    }
}

@Composable
private fun PresenceBubble(kind: MiaPresenceKind) {
    val shape = bubbleShape(isUser = false)
    Box(
        modifier = Modifier
            .height(36.dp)
            .shadow(2.dp, shape)
            .background(MiaColors.MiaBubble, shape)
            .padding(
                horizontal = if (kind == MiaPresenceKind.Typing) 14.dp else 16.dp,
                vertical = if (kind == MiaPresenceKind.Typing) 10.dp else 9.dp,
            ),
        contentAlignment = Alignment.Center,
    ) {
        when (kind) {
            MiaPresenceKind.Typing -> TypingDots()
            MiaPresenceKind.Recording -> RecordingMicPulse()
        }
    }
}

@Composable
private fun TypingDots() {
    val anim = rememberInfiniteTransition(label = "typing")
    val t by anim.animateFloat(
        0f,
        1f,
        infiniteRepeatable(tween(1200, easing = LinearEasing), RepeatMode.Restart),
        label = "t",
    )
    Row(verticalAlignment = Alignment.Bottom) {
        repeat(3) { i ->
            val phase = t * 2f * Math.PI.toFloat() - i * 0.75f
            val y = sin(phase) * 4f
            val alpha = 0.55f + 0.35f * ((sin(phase) + 1f) / 2f)
            if (i > 0) Box(Modifier.width(4.dp))
            Box(
                Modifier
                    .size(6.dp)
                    .clip(CircleShape)
                    .background(MiaColors.TextMuted.copy(alpha = alpha)),
            )
        }
    }
}

@Composable
private fun RecordingMicPulse() {
    val anim = rememberInfiniteTransition(label = "mic")
    val pulse by anim.animateFloat(
        0.75f,
        1f,
        infiniteRepeatable(tween(900), RepeatMode.Reverse),
        label = "pulse",
    )
    Icon(
        Icons.Rounded.Mic,
        contentDescription = null,
        tint = Color(0xFFE54D42),
        modifier = Modifier
            .size(18.dp)
            .alpha(pulse)
            .scale(0.92f + pulse * 0.08f),
    )
}
